python/cudf/cudf/core/frame.py

```diff
@@ -1723,6 +1723,105 @@ def __arrow_array__(self, type=None):
             "consider using .to_arrow()"
         )
 
+    def round(self, decimals=0):
+        """
+        Round a DataFrame to a variable number of decimal places.
+
+        Parameters
+        ----------
+        decimals : int, dict, Series
+            Number of decimal places to round each column to. If an int is
+            given, round each column to the same number of places.
+            Otherwise dict and Series round to variable numbers of places.
+            Column names should be in the keys if `decimals` is a
+            dict-like, or in the index if `decimals` is a Series. Any
+            columns not included in `decimals` will be left as is. Elements
+            of `decimals` which are not columns of the input will be
+            ignored.
+
+        Returns
+        -------
+        DataFrame
+            A DataFrame with the affected columns rounded to the specified
+            number of decimal places.
+
+        Examples
+        --------
+        >>> df = cudf.DataFrame(
+                [(.21, .32), (.01, .67), (.66, .03), (.21, .18)],
+        ...     columns=['dogs', 'cats']
+        ... )
+        >>> df
+            dogs  cats
+        0  0.21  0.32
+        1  0.01  0.67
+        2  0.66  0.03
+        3  0.21  0.18
+
+        By providing an integer each column is rounded to the same number
+        of decimal places
+
+        >>> df.round(1)
+            dogs  cats
+        0   0.2   0.3
+        1   0.0   0.7
+        2   0.7   0.0
+        3   0.2   0.2
+
+        With a dict, the number of places for specific columns can be
+        specified with the column names as key and the number of decimal
+        places as value
+
+        >>> df.round({'dogs': 1, 'cats': 0})
+            dogs  cats
+        0   0.2   0.0
+        1   0.0   1.0
+        2   0.7   0.0
+        3   0.2   0.0
+
+        Using a Series, the number of places for specific columns can be
+        specified with the column names as index and the number of
+        decimal places as value
+
+        >>> decimals = cudf.Series([0, 1], index=['cats', 'dogs'])
+        >>> df.round(decimals)
+            dogs  cats
+        0   0.2   0.0
+        1   0.0   1.0
+        2   0.7   0.0
+        3   0.2   0.0
+        """
+
+        def _series_round(s, decimals):
+            if pd.api.types.is_numeric_dtype(s.dtype):
+                return s.round(decimals)
+            return s
+
+        if isinstance(decimals, (dict, cudf.Series, pd.Series)):
+            if (
+                isinstance(decimals, (cudf.Series, pd.Series))
+                and not decimals.index.is_unique
+            ):
+                raise ValueError("Index of decimals must be unique")
+            new_cols = []
+            for c in self.columns:
+                if c in decimals.keys():
+                    new_cols.append(_series_round(self[c], decimals[c]))
+                else:
+                    new_cols.append(self[c])
```

Comment:
In cuDF, `DataFrame.columns` is non-trivial to evaluate (look at the `columns` property of `DataFrame` to get an idea). If you want to loop over the columns of a DataFrame and apply some operation on each one, the following pattern is preferred:

```python
for col in self._data.values():
    # operate on `col`  # note `col` is a `Column` rather than a `Series`
```

You can read about the `_data` attribute [here](https://docs.rapids.ai/api/cudf/stable/internals.html#columnaccessor)

Reply:
And take care of shallow copy and deep copy in this scenario.